src/systems/filecoin_blockchain/chainsync/_index.md

```diff
@@ -1,58 +1,490 @@
+- **Chain State and Finality**:
+  - In this state, the **chain MUST NOT advance** beyond whatever the node already has:
+    - `FinalityTipset` does not change.
+    - No new blocks are reported to consumers/users of `ChainSync` yet.
+    - The chain state provided is the available `Blocks` and `StateTree` for all available epochs,
+      specially the `FinalityTipset`.
+    - finality must not move forward here because there are serious attack vectors where a node can be forced to end up on the wrong fork if finality advances before validation is complete up to the block production fringe.
+  - Validation must advance, all the way to the block production fringe:
+    - Validate the whole chain, from `FinalityTipset` to `BestTargetHead`
+    - The node can reach `BestTargetHead` only to find out it was invalid, then has to update `BestTargetHead` with next best one, and sync to it
+      (without having advanced `FinalityTipset` yet, as otherwise we may end up on the wrong fork)
+- **security conditions to transition out:**
+  - Gaps between `ChainSync.FinalityTipset ... ChainSync.BestTargetHead` have been closed:
+    - All `Blocks` and their content MUST be fetched, stored, linked, and validated locally.
+      This includes `BlockHeaders`, `Messages`, etc.
+    - Bad heads have been expunged from `ChainSync.TargetHeads`. Bad heads include heads that initially
+      seemed good but turned out invalid, or heads that `ChainSync` has failed to connect (ie. cannot
+      fetch ancestors connecting back to `ChainSync.FinalityTipset` within a reasonable amount of time).
+    - All blocks between `ChainSync.FinalityTipset ... ChainSync.TargetHeads` have been validated
+      This means all blocks _before_ the best heads.
+  - Not under a temporary network partition
+- **transitions out:**
+  - once gaps between `ChainSync.FinalityTipset ... ChainSync.TargetHeads` are closed: move to `CHAIN_FOLLOW`
+  - (Perhaps moving to `CHAIN_FOLLOW` when 1-2 blocks back in validation may be ok.
+    - we dont know we have the right head until we validate it, so if other heads of similar height are right/better, we wont know till then.)
+
+## ChainSync FSM: `CHAIN_FOLLOW`
+
+- While in theis state:
+  - `ChainSync` is well-bootstrapped, and has an initial **trusted** `StateTree` to start from.
+  - `ChainSync` fetches and validates blocks (see _Block Fetching and Validation_).
+  - `ChainSync` is receiving and validating latest `Blocks` from `BlockPubsub`
+  - `ChainSync` DOES NOT have unvalidated blocks between `ChainSync.FinalityTipset` and `ChainSync.TargetHeads`
+  - `ChainSync` MUST drop back to another state if security conditions change.
+  - Keep a set of gap measures:
+    - `BlockGap` is the number of remaining blocks to validate between the Validated blocks and `BestTargetHead`.
+      - (ie how many epochs do we need to validate to have validated `BestTargetHead`. does not include null blocks)
```

Comment:
I'm not sure I understand the BlockGap purpose here. The idea being stop doing any progress work and just do validation if you fall behind? If so makes sense. In which case it should be a multiple of expectedNumberOfBlocksPerRound (and in that sense a proxy for epochGap in some way)

Reply:
> -  The idea being stop doing any progress work and just do validation if you fall behind? 

yes.

- and: the difference between `BlockGap` and `EpochGap` is null blocks. we can tolerate more of those than real blocks gap.

Reply:
Fair point